### PR TITLE
Optimize SiLU (Swish) op in PyTorch

### PR DESCRIPTION
**Pull request by xiaomengy**
Summary:
Optimize SiLU (Swish) op in PyTorch.

Some benchmark result

input = torch.rand(1024, 32768, dtype=torch.float, device="cpu")
forward: 221ms -> 133ms
backward: 600ms -> 170ms

input = torch.rand(1024, 32768, dtype=torch.double, device="cpu")
forward: 479ms -> 297ms
backward: 1438ms -> 387ms

input = torch.rand(8192, 32768, dtype=torch.float, device="cuda")
forward: 24.34ms -> 9.83ms
backward: 97.05ms -> 29.03ms

input = torch.rand(4096, 32768, dtype=torch.double, device="cuda")
forward: 44.24ms -> 30.15ms
backward: 126.21ms -> 49.68ms

Test Plan: buck test mode/dev-nosan //caffe2/test:nn -- "SiLU"

Differential Revision: D23093593

### CODE REVIEW

**Comment by BIT-silence** [comment]
linke https://github.com/pytorch/pytorch/issues/43017

File: aten/src/ATen/native/Activation.cpp [hunk 1]
```diff
@@ -194,21 +198,64 @@ Tensor & celu_(Tensor & self, Scalar alpha) {
...
+    Tensor& buffer,
+    const Tensor& input) {
+  auto iter = TensorIteratorConfig()
+                  .set_check_mem_overlap(true)
+                  .add_output(result)
+                  .add_output(buffer)
+                  .add_input(input)
+                  .build();
+  silu_stub(iter.device_type(), iter);
+  return std::forward_as_tuple(result, buffer);
```
Comment (ngimel) [inline aten/src/ATen/native/Activation.cpp:219]: at least on cuda, since you have to read `x` in backward anyway, recomputing `sigmoid` is pretty cheap (compared to writing it in forward and reading it in backward), so given that it also requires extra memory, maybe you can rewrite it without buffer?

Comment (BIT-silence) [reply aten/src/ATen/native/Activation.cpp:219]: Yes, this is a good point, actually I am testing this and trying to find a proper way to do that on CUDA.

File: aten/src/ATen/native/cpu/Activation.cpp [hunk 2]
```diff
@@ -589,6 +596,111 @@ void glu_backward_kernel(TensorIterator& iter) {
...
+          cpu_kernel_vec(
+              iter2,
+              [](scalar_t a, scalar_t b) { return a * b; },
+              [](Vec256<scalar_t> a, Vec256<scalar_t> b) { return a * b; });
+        }
+      });
+}
+
+template <typename T>
+void SiLUBackwardContiguousKernel(TensorIterator* it) {
```
Comment (gchanan) [inline aten/src/ATen/native/cpu/Activation.cpp:666]: why do some kernels take TensorIterators as references and some as points?

File: aten/src/ATen/native/Activation.cpp (hunk 1 above)
Comment (BIT-silence) [reply aten/src/ATen/native/Activation.cpp]: From my testing, removing the buffer will make cpu backward slightly slower but forward much faster. For CUDA side, both forward and backward become faster. I have updated the code as well as the benchmark results.

File: aten/src/ATen/native/cpu/Activation.cpp (hunk 2 above)
Comment (BIT-silence) [reply aten/src/ATen/native/cpu/Activation.cpp]: Removed.

File: aten/src/ATen/native/Activation.cpp [hunk 3]
```diff
@@ -195,20 +199,37 @@ Tensor & celu_(Tensor & self, Scalar alpha) {
...
 }
 
 Tensor& silu_(Tensor& self) {
-  return self.mul_(at::sigmoid(self));
+  return at::silu_out(self, self);
 }
 
 Tensor& silu_out(Tensor& result, const Tensor& self) {
-  return at::mul_out(result, self, at::sigmoid(self));
+  auto iter = TensorIterator::unary_op(result, self);
```
Comment (ngimel) [inline aten/src/ATen/native/Activation.cpp:212]: please check that result type is same as self, so that `out` variant of the function properly errors out if user sent a wrong argument

File: aten/src/ATen/native/Activation.cpp [hunk 4]
```diff
@@ -1,5 +1,7 @@
 #include <ATen/native/Activation.h>
 
+#include <tuple>
```
Comment (ngimel) [inline aten/src/ATen/native/Activation.cpp:3]: do you need `tuple` include here?

File: test/test_torch.py [hunk 5]
```diff
@@ -16725,20 +16725,32 @@ def test_hardsigmoid(self, device, dtype):
 
     @dtypes(torch.float, torch.double)
     def test_silu(self, device, dtype):
-        inputValues = [-1000, -1, 0, 0.5, 1, 2, 1000]
-        expectedOutput = [0.0000, -0.2689, 0, 0.3112, 0.7312, 1.7616, 1000]
-        precision_4dps = 0.0002
+        input_np = np.random.randn(5, 8)
```
Comment (ngimel) [inline test/test_torch.py:16728]: I think it's better to just add silu to TestTorchMathOps (you can easily create np reference lambda for it) and remove this test altogether. If you don't want to do that, you can use `compare_with_numpy` helper function here to handle comparison.

File: aten/src/ATen/native/native_functions.yaml [hunk 6]
```diff
@@ -2659,6 +2659,9 @@
   use_c10_dispatcher: full
   python_module: nn
 
+- func: silu_backward.grad_input(Tensor grad_output, Tensor self, *, Tensor(a!) grad_input) -> Tensor(a!)
```
Comment (ngimel) [inline aten/src/ATen/native/native_functions.yaml:2662]: do you need to expose this overload?

File: tools/autograd/templates/Functions.cpp [hunk 7]
```diff
@@ -959,6 +959,13 @@ Tensor infinitely_differentiable_gelu_backward(
   return cdf.addcmul_(self, pdf, kAlpha).mul_(grad);
 }
 
+Tensor infinitely_differentiable_silu_backward(
+    const Tensor& grad_output,
+    const Tensor& input) {
```
Comment (ngimel) [inline tools/autograd/templates/Functions.cpp:964]: yep, without buffer it looks much better :-)

File: torch/csrc/api/include/torch/nn/modules/activation.h [hunk 8]
```diff
@@ -567,6 +567,27 @@ class TORCH_API GELUImpl : public torch::nn::Cloneable<GELUImpl> {
 /// module storage semantics.
 TORCH_MODULE(GELU);
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ SiLU ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+/// Applies gelu over a given input.
```
Comment (ngimel) [inline torch/csrc/api/include/torch/nn/modules/activation.h:572]: silu

**Review by ngimel — COMMENTED** [review]
This looks great, thank you!

File: aten/src/ATen/native/Activation.cpp (hunk 3 above)
Comment (BIT-silence) [reply aten/src/ATen/native/Activation.cpp:213]: Done

File: test/test_torch.py (hunk 5 above)
Comment (BIT-silence) [reply test/test_torch.py:16800]: The issue is silu is a function in torch.nn.functional, not an attribute in tensor. It seems currently TestTorchMathMeta test the methods by getattr(x, opstr). Not sure if there is some proper way to do that.

File: aten/src/ATen/native/Activation.cpp (hunk 4 above)
Comment (BIT-silence) [reply aten/src/ATen/native/Activation.cpp]: Removed.

File: aten/src/ATen/native/native_functions.yaml (hunk 6 above)
Comment (BIT-silence) [reply aten/src/ATen/native/native_functions.yaml]: Removed.

File: aten/src/ATen/native/Activation.cpp [hunk 9]
```diff
@@ -195,20 +199,37 @@ Tensor & celu_(Tensor & self, Scalar alpha) {
...
 }
 
-Tensor silu_backward(const Tensor& grad, const Tensor& self) {
-  auto self_sigmoid = at::sigmoid(self);
-  return grad * (self_sigmoid * (1 + self * (1 - self_sigmoid)));
+Tensor silu_backward(
+    const Tensor& grad_output,
+    const Tensor& input) {
+  Tensor grad_input =
+      at::native::empty_like(input, at::MemoryFormat::Contiguous);
```
[the remark this answers is not in the document]
Comment (BIT-silence) [reply aten/src/ATen/native/Activation.cpp]: Done.

File: aten/src/ATen/native/Activation.cpp [hunk 10]
```diff
@@ -195,20 +199,37 @@ Tensor & celu_(Tensor & self, Scalar alpha) {
 }
 
 Tensor silu(const Tensor& self) {
-  return self * at::sigmoid(self);
+  Tensor result = at::native::empty_like(self, at::MemoryFormat::Contiguous);
```
[the remark this answers is not in the document]
Comment (BIT-silence) [reply aten/src/ATen/native/Activation.cpp]: Done.

File: torch/csrc/api/include/torch/nn/modules/activation.h (hunk 8 above)
Comment (BIT-silence) [reply torch/csrc/api/include/torch/nn/modules/activation.h]: Done.

File: test/test_torch.py (hunk 5 above)
Comment (ngimel) [reply test/test_torch.py:16800]: Yeah, ok, nevermind then.

File: test/test_torch.py [hunk 11]
```diff
@@ -16797,20 +16797,32 @@ def test_hardsigmoid(self, device, dtype):
 
     @dtypes(torch.float, torch.double)
     def test_silu(self, device, dtype):
```
Comment (ngimel) [inline test/test_torch.py:16799]: you need @skipIfNoScipy here

**Review by ngimel — APPROVED** [review]

File: test/test_torch.py (hunk 11 above)
Comment (BIT-silence) [reply test/test_torch.py:16800]: Done.
